File: app/src/main/java/com/scottmangiapane/cardclicks/game/Game.java
package com.scottmangiapane.cardclicks.game;

import android.util.Log;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Locale;
import java.util.Random;

public class Game {
    private final Card[] current;
    private final Card[] remaining;
    private final Random generator = new Random();

    public Game(int size) {
        List<Card> deck = new ArrayList<>(Arrays.asList(Deck.DECK));
        Collections.shuffle(deck);
        current = deck.subList(0, size).toArray(new Card[0]);
        remaining = deck.subList(size, deck.size()).toArray(new Card[0]);
        forceSet();
    }

    private void forceSet(int... indexes) {
        if (getSets().isEmpty()) {
            int swapIndex = generator.nextInt(current.length);
            if (indexes.length > 0) {
                swapIndex = indexes[generator.nextInt(indexes.length)];
            }
            int i1, i2;
            do {
                i1 = generator.nextInt(current.length);
            } while (i1 == swapIndex);
            do {
                i2 = generator.nextInt(current.length);
            } while (i2 == swapIndex || i2 == i1);
            Card newCard = Deck.getComplimentaryCard(current[i1], current[i2]);
            setCard(swapIndex, newCard);
            Log.d("Game", String.format(
                    Locale.getDefault(),
                    "Forced set with cards [%d, %d, %d]",
                    swapIndex, i1, i2));
        }
    }

    private void setCard(int index, Card card) {
        int swapIndex = Arrays.asList(remaining).indexOf(card);
        Card temp = current[index];
        current[index] = remaining[swapIndex];
        remaining[swapIndex] = temp;
    }

    public Card get(int index) {
        return current[index];
    }

    public List<Integer[]> getSets() {
        List<Integer[]> sets = new ArrayList<>();
        for (int i1 = 0; i1 < current.length - 2; i1++) {
            for (int i2 = i1 + 1; i2 < current.length - 1; i2++) {
                for (int i3 = i2 + 1; i3 < current.length; i3++) {
                    if (Deck.isSet(current[i1], current[i2], current[i3])) {
                        Log.d("Game", String.format(
                                Locale.getDefault(),
                                "Found set with cards [%d, %d, %d]",
                                i1, i2, i3));
                        Integer[] set = new Integer[]{i1, i2, i3};
                        sets.add(set);
                    }
                }
            }
        }
        return sets;
    }

    public boolean isSet(int i1, int i2, int i3) {
        return Deck.isSet(current[i1], current[i2], current[i3]);
    }

    public void rotateCards(int... indexes) {
        shuffleArray(remaining);
        for (int i = 0; i < indexes.length; i++) {
            setCard(indexes[i], remaining[i]);
        }
        forceSet(indexes);
    }

    private static <T> void shuffleArray(T[] array) {
        int index;
        T temp;
        Random random = new Random();
        for (int i = array.length - 1; i > 0; i--)
        {
            index = random.nextInt(i + 1);
            temp = array[index];
            array[index] = array[i];
            array[i] = temp;
        }
    }

    public int size() {
        return current.length;
    }
}

File: app/src/main/java/com/scottmangiapane/cardclicks/game/Game.kt
package com.scottmangiapane.cardclicks.game

import android.util.Log
import java.util.Locale
import java.util.Random

class Game(size: Int) {
    private val current: Array<Card>
    private val remaining: Array<Card>
    private val generator = Random()

    init {
        val deck = Deck.DECK.toMutableList()
        deck.shuffle()
        current = deck.subList(0, size).toTypedArray()
        remaining = deck.subList(size, deck.size).toTypedArray()
        forceSet()
    }

    private fun forceSet(vararg indexes: Int) {
        if (getSets().isEmpty()) {
            var swapIndex = generator.nextInt(current.size)
            if (indexes.isNotEmpty()) swapIndex = indexes[generator.nextInt(indexes.size)]
            var i1: Int
            var i2: Int
            do {
                i1 = generator.nextInt(current.size)
            } while (i1 == swapIndex)
            do {
                i2 = generator.nextInt(current.size)
            } while (i2 == swapIndex || i2 == i1)
            val newCard = Deck.getComplimentaryCard(current[i1], current[i2])
            setCard(swapIndex, newCard)
            Log.d(
                "Game", String.format(
                    Locale.getDefault(),
                    "Forced set with cards [%d, %d, %d]",
                    swapIndex, i1, i2
                )
            )
        }
    }

    private fun setCard(index: Int, card: Card) {
        val swapIndex = remaining.indexOf(card)
        val temp = current[index]
        current[index] = remaining[swapIndex]
        remaining[swapIndex] = temp
    }

    operator fun get(index: Int): Card {
        return current[index]
    }

    fun getSets(): List<Array<Int>> {
        val sets = mutableListOf<Array<Int>>()
        for (i1 in 0..<current.size - 2) {
            for (i2 in i1 + 1..<current.size - 1) {
                for (i3 in i2 + 1..<current.size) {
                    if (Deck.isSet(current[i1], current[i2], current[i3])) {
                        Log.d(
                            "Game", String.format(
                                Locale.getDefault(),
                                "Found set with cards [%d, %d, %d]",
                                i1, i2, i3
                            )
                        )
                        sets.add(arrayOf(i1, i2, i3))
                    }
                }
            }
        }
        return sets
    }

    fun isSet(i1: Int, i2: Int, i3: Int): Boolean {
        return Deck.isSet(current[i1], current[i2], current[i3])
    }

    fun rotateCards(vararg indexes: Int) {
        shuffleArray(remaining)
        for (i in indexes.indices) setCard(indexes[i], remaining[i])
        forceSet(*indexes)
    }

    companion object {
        private fun <T> shuffleArray(array: Array<T>) {
            val random = Random()
            for (i in array.size - 1 downTo 1) {
                val index = random.nextInt(i + 1)
                val temp = array[index]
                array[index] = array[i]
                array[i] = temp
            }
        }
    }
}
